src/base/ActionsRouter.sol

```diff
@@ -0,0 +1,143 @@
+                else if (action == Actions.BURN_POSITION) _burnPosition(params[actionIndex]);
+                else if (action == Actions.DONATE) _donate(params[actionIndex]);
+                else _handleAdditionalActions2(action, params[actionIndex]);
+            } else {
+                if (action == Actions.SETTLE) _settle(params[actionIndex]);
+                else if (action == Actions.TAKE) _take(params[actionIndex]);
+                else if (action == Actions.MINT_6909) _mint6909(params[actionIndex]);
+                else if (action == Actions.BURN_6909) _burn6909(params[actionIndex]);
+                else if (action == Actions.CLEAR_DELTA) _clearDelta(params[actionIndex]);
+                else if (action == Actions.SWEEP_ETH) _sweepETH(params[actionIndex]);
+                else _handleAdditionalActions2(action, params[actionIndex]);
+            }
+        }
+
+        // TODO do we want to return anything?
+    }
+
+    /// @notice function that returns the msg.sender who called executeActions
+    function _msgSender() internal view virtual returns (address) {
+        return Locker.get();
+    }
+
+    /// @notice function to settle the open delta of a currency
+    /// @dev The `_pay` function must implement necessary safety checks on the value of payer.
+    /// If approvals are used on this contract, do not allow `_pay` to pull tokens from any payer.
+    /// @dev The function is virtual and can be overrided if different parameters or logic are required.
+    function _settle(bytes memory params) internal virtual {
+        // TODO decode in calldata
+        (Currency currency, address payer) = abi.decode(params, (Currency, address));
+
+        int256 delta = poolManager.currencyDelta(address(this), currency);
+        if (delta > 0) revert();
+
+        if (currency.isNative()) {
+            poolManager.settle{value: uint256(-delta)}();
+        } else {
+            poolManager.sync(currency);
+            _pay(currency, payer, address(poolManager), uint256(-delta));
+            poolManager.settle();
+        }
+    }
+
+    /// @notice abstract function to implement the payment of tokens to the pool manager during a settle
+    function _pay(Currency currency, address payer, address recipient, uint256 amount) internal virtual;
+
+    /// @notice function to take owed currency from the pool manager to a recipient
+    /// @dev The function is virtual and can be overrided if different parameters or logic are required.
+    function _take(bytes memory params) internal virtual {
+        // TODO decode in calldata
+        (Currency currency, address recipient) = abi.decode(params, (Currency, address));
+
+        int256 delta = poolManager.currencyDelta(address(this), currency);
+        if (delta < 0) revert();
+
+        poolManager.take(currency, recipient, uint256(delta));
+    }
+
+    /// @notice function to sweep any excess ETH back to a recipient
+    function _sweepETH(bytes memory params) internal virtual {
```

Comment:
do we need a sweep function for tokens?

Reply:
yeah i did wonder that... i assumed the only reason people will ever handle tokens are for a final `settle` or `take` on their actions and that tokens should never need to be in the contract? But if you can think of an example i could add it

Reply:
someone could send tokens to the contract by accident

Reply:
But these functions dont need to be part of the "Actions"